### PR TITLE
Implement the Crop Tool for artboard resizing

### LINKED ISSUE #493: Implement the Crop Tool for resizing the artboard
Right now it should be fine to just drag and draw a rectangle with the Crop Tool which becomes the new artboard bounds. <kbd>Shift</kbd> to constrain to a square. Bonus points if this can be used to create multiple artboards. When the viewport's view of the canvas is rotated, the artboard remains axis-aligned in document space, meaning it stays tilted in viewport space (unlike, say, a rectangle if you decided to draw one with the Rectangle Tool). Additional bonus points if we can have little square blue drag handles overlaid at the corners and edge-centers of the artboard so dragging those would perform resizing along those edges (no rotation though). The resulting document bounds (width, height, X position, and Y position) should each be rounded to integer values in document space.

### PR DESCRIPTION
**Pull request by 0HyperCube**
Closes #493

TODO:

- [x] Size to square with shift and centre with alt
- [x] Snap artboards & layers to other artboards
- [x] Fix resize behaviour with a rotated canvas - we want to calculate the bounding box with the shape transformed with just that layer's transform, and then transform the resulting box.
- [x] Allow artboards to be dragged and moved.

### CODE REVIEW

File: editor/src/document/artboard_message_handler.rs [hunk 1]
```diff
@@ -73,6 +73,20 @@ impl MessageHandler<ArtboardMessage, ()> for ArtboardMessageHandler {
...
 				}
 			}
+			ResizeArtboard { artboard, position, size } => {
+				responses.push_back(
+					ArtboardMessage::DispatchOperation(
+						DocumentOperation::SetLayerTransform {
+							path: artboard,
+							transform: DAffine2::from_scale_angle_translation(size.into(), 0., position.into()).to_cols_array(),
+						}
+						.into(),
```
Comment (TrueDoctor) [inline editor/src/document/artboard_message_handler.rs:83]: is this into even required here?

File: editor/src/document/document_message_handler.rs [hunk 2]
```diff
@@ -226,6 +230,22 @@ impl DocumentMessageHandler {
...
+				.filter_map(|path| self.graphene_document.viewport_bounding_box(path).ok()?)
+				.collect(),
+			self.artboard_message_handler
+				.artboard_ids
+				.iter()
+				.filter(|id| Some(**id) != ignore_artboard)
+				.filter_map(|path| self.artboard_message_handler.artboards_graphene_document.viewport_bounding_box(&[*path]).ok()?)
+				.collect::<Vec<[DVec2; 2]>>(),
+		]
+		.concat()
```
Comment (TrueDoctor) [inline editor/src/document/document_message_handler.rs:246]: would it be possible to return a chained iterator instead?

File: editor/src/viewport_tools/tools/crop.rs [hunk 3]
```diff
@@ -1,30 +1,334 @@
...
+	DocumentIsDirty,
 
 	// Tool-specific messages
-	MouseMove,
+	MouseDown,
+	MouseMove {
+		axis_align: Key,
+		centre: Key,
+	},
+	MouseUp,
```
Comment (TrueDoctor) [inline editor/src/viewport_tools/tools/crop.rs:40]: We should probably use the more general Pointer move here, the usage kinda differs throughout he codebase buf iirc we agreed that it should be PointerMove etc

File: editor/src/document/artboard_message_handler.rs (hunk 1 above)
Comment (0HyperCube) [reply editor/src/document/artboard_message_handler.rs:83]: @TrueDoctor It boxes the operation.

Comment (0HyperCube) [reply editor/src/document/artboard_message_handler.rs:83]: I have converted it to a manual `Box::new` to add clarity

File: editor/src/document/document_message_handler.rs (hunk 2 above)
Comment (0HyperCube) [reply editor/src/document/document_message_handler.rs:246]: Sure - forgot about that :)

File: editor/src/document/artboard_message_handler.rs (hunk 1 above)
Comment (TrueDoctor) [reply editor/src/document/artboard_message_handler.rs:83]: Ah, right, always forget that

File: editor/src/viewport_tools/tools/crop.rs (hunk 3 above)
Comment (0HyperCube) [reply editor/src/viewport_tools/tools/crop.rs:40]: I have cleaned up in the rest of the codebase.

Comment (TrueDoctor) [reply editor/src/viewport_tools/tools/crop.rs]: Awesome!

**Comment by Keavon** [comment]
Giving my approval because I dig it! Feature works splendidly and is very production-ready. However I suggest we have someone review the code before merging if there is time. I won't have time to review code before alpha.

**Review by Keavon — APPROVED** [review]

**Review by TrueDoctor — APPROVED** [review]
As far as I can tell this all looks good!

**Pull request merged**